src/nvim/path.c

```diff
@@ -597,45 +597,68 @@ static size_t do_path_expand(garray_T *gap, const char_u *path,
   *s = NUL;
 
   Directory dir;
-  // open the directory for scanning
-  if (os_scandir(&dir, *buf == NUL ? "." : (char *)buf)) {
+  if (os_isdir(*buf == NUL ? (char_u *)"." : (char_u *)buf)) {
     // Find all matching entries.
-    char_u *name;
-    while((name = (char_u *) os_scandir_next(&dir))) {
-      if ((name[0] != '.' || starts_with_dot)
-          && ((regmatch.regprog != NULL && vim_regexec(&regmatch, name, 0))
-              || ((flags & EW_NOTWILD)
-                  && fnamencmp(path + (s - buf), name, e - s) == 0))) {
-        STRCPY(s, name);
-        len = STRLEN(buf);
-
-        if (starstar && stardepth < 100) {
-          /* For "**" in the pattern first go deeper in the tree to
-           * find matches. */
-          STRCPY(buf + len, "/**");
-          STRCPY(buf + len + 3, path_end);
-          ++stardepth;
-          (void)do_path_expand(gap, buf, len + 1, flags, true);
-          --stardepth;
-        }
 
-        STRCPY(buf + len, path_end);
-        if (path_has_exp_wildcard(path_end)) {      /* handle more wildcards */
-          /* need to expand another component of the path */
-          /* remove backslashes for the remaining components only */
-          (void)do_path_expand(gap, buf, len + 1, flags, false);
-        } else {
-          /* no more wildcards, check if there is a match */
-          /* remove backslashes for the remaining components only */
-          if (*path_end != NUL)
-            backslash_halve(buf + len + 1);
-          if (os_file_exists(buf)) {          /* add existing file */
-            addfile(gap, buf, flags);
-          }
-        }
+    // check if the directory is empty
+    // Note: this updates 'dir', so we must do it outside the loop
+    bool dir_empty = !os_scandir(&dir, *buf == NUL ? "." : (char *)buf);
+    char_u *name;
+    bool more_entries = true;
+    int entries = 0;
+    while(more_entries) {
+      // first two loops are used to check if . and .. match
+      switch(entries) {
```

Comment:
This is a lot more roundabout than I was hoping for. How about making a static function that wraps `os_scandir_next` and always returns `.` and `..` for the first two calls?